packages/ra-keycloak/src/authProvider.ts

```diff
@@ -103,55 +77,147 @@ export type PermissionsFunction = (decoded: KeycloakTokenParsed) => any;
-            ? Promise.resolve()
-            : Promise.reject('Failed to obtain access token.');
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (keycloakClient.authenticated && keycloakClient.token) {
+            return;
+        }
+        // not authenticated: save the location that the user tried to access
+        localStorage.setItem(
+            PreviousLocationStorageKey,
+            window.location.href.replace(window.location.origin, '')
+        );
+        throw new Error('Failed to obtain access token.');
     },
     async getPermissions() {
-        if (!client.token) {
-            return Promise.resolve(false);
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (!keycloakClient.token) {
+            return false;
         }
-        const decoded = jwt_decode<KeycloakTokenParsed>(client.token);
-        return Promise.resolve(
-            options.onPermissions ? options.onPermissions(decoded) : decoded
-        );
+        const decoded = jwt_decode<KeycloakTokenParsed>(keycloakClient.token);
+        return options.onPermissions ? options.onPermissions(decoded) : decoded;
     },
     async getIdentity() {
-        if (client.token) {
-            const decoded = jwt_decode<KeycloakTokenParsed>(client.token);
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+        if (keycloakClient.token) {
+            const decoded = jwt_decode<KeycloakTokenParsed>(
+                keycloakClient.token
+            );
             const id = decoded.sub || '';
             const fullName = decoded.preferred_username;
-            return Promise.resolve({ id, fullName });
+            return { id, fullName };
+        }
+        throw new Error('Failed to get identity.');
+    },
+    async handleCallback(params: any) {
+        await initKeyCloakClient(keycloakClient, options.initOptions);
+        await isAuthenticated(keycloakClient);
+
+        if (!keycloakClient.authenticated) {
+            throw new Error('Failed to obtain access token.');
         }
-        return Promise.reject('Failed to get identity.');
     },
 });
+
+/**
+ * It seems the Keycloak init function may initially return before having authenticated the user.
+ * To ensure we have the correct state, we need to wait for the onAuthSuccess event.
+ */
+const isAuthenticated = (keycloakClient: Keycloak) => {
+    return new Promise((resolve, reject) => {
+        let authenticated = false;
+        keycloakClient.onAuthSuccess = () => {
+            resolve(true);
+            keycloakClient.onAuthSuccess = null;
+        };
+        setTimeout(() => {
+            if (!authenticated) {
+                resolve(false);
+                keycloakClient.onAuthSuccess = null;
+            }
+        }, 2000);
```

Comment:
`2000`: this hard-coded timeout feels a bit short to me (in case of slow internet connection for instance). Shouldn't we increase it, or make it configurable?